### PR TITLE
Feat/inbox tests

### PR DESCRIPTION
**Pull request by sayaansaha**
Fixes # Improved unit test coverage of LeanplumInbox.java to 100% method coverage and 80% lines covered

Improved unit test coverage of LeanplumInboxMessage.java  to 90% of methods covered and 82% lines covered

## Proposed Changes
  - Added method  enableInboxImagePrefetching()

### CODE REVIEW

**Comment by sayaansaha** [comment]
build failed due to jenkins error "error: branch 'pr-164' not found"

**Comment by Enuviel** [comment]
test this please

**Comment by Enuviel** [comment]
@sayaansaha   next time please fix a Jenkins issue and run test again with a comment "test this please", instead of comment why it fails. Thank you!

File: AndroidSDKCore/src/main/java/com/leanplum/LeanplumInbox.java [hunk 1]
```diff
@@ -85,6 +85,13 @@ public static void disableImagePrefetching() {
     isInboxImagePrefetchingEnabled = false;
   }
 
+  /**
+   * Enable prefetching images.
+   */
+  public static void enableInboxImagePrefetching() {
```
Comment (Enuviel) [inline AndroidSDKCore/src/main/java/com/leanplum/LeanplumInbox.java]: You shouldn't change original class when you write tests. Please remove this method.

Comment (sayaansaha) [reply AndroidSDKCore/src/main/java/com/leanplum/LeanplumInbox.java]: I disagree, While writing the tests I discovered that calling disableImagePrefetching can not be undone, It makes a lot of sense for this method to be there and should have been created in the first place.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java [hunk 2]
```diff
@@ -48,141 +49,191 @@
...
+  }
+
+  @Test
+  public void testDisablePrefetching() {
+    leanplumInbox.disableImagePrefetching();
+    assertFalse(leanplumInbox.isInboxImagePrefetchingEnabled());
+  }
+
+  @Test
+  public void testInvalidMessageIdIsRejected() {
```
Comment (Enuviel) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java]: Why is this one a separate test? Please put old lines 185-186 back and if you want, please put comment what those lines do.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java [hunk 3]
```diff
@@ -48,141 +49,191 @@
...
+    map.put(Constants.Keys.DELIVERY_TIMESTAMP, delivery.getTime());
+    map.put(Constants.Keys.EXPIRATION_TIMESTAMP, expiration.getTime());
+    map.put(Constants.Keys.IS_READ, true);
+
+    LeanplumInboxMessage invalidMessage = LeanplumInboxMessage.createFromJsonMap("messageId", map);
+    assertNull(invalidMessage);
+  }
+
+  @Test
+  public void testRemoveMessageFromInbox() {
```
Comment (Enuviel) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java]: why is this a separate method? you have it in testInbox (new lines 144-145)

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java [hunk 4]
```diff
@@ -48,141 +49,191 @@
...
+
+    assertEquals(1, leanplumInbox.count());
+
+    leanplumInbox.downloadMessages();
+
+    assertEquals(2, leanplumInbox.count());
+  }
+
+  @Test
+  public void testGettingUnreadMessagesAfterARead() {
```
Comment (Enuviel) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java]: is this test implemented here  https://github.com/Leanplum/Leanplum-Android-SDK/pull/164/files#diff-d4fac6525f945e4a092843223e65630dR132?

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    map.put(Constants.Keys.EXPIRATION_TIMESTAMP, expiration.getTime());
+    map.put(Constants.Keys.IS_READ, false);
+    LeanplumInboxMessage message = LeanplumInboxMessage
+        .createFromJsonMap("messageId##00", map);
+
+    assertEquals(false, message.isRead());
+
+    message.read();
+
+    assertEquals(true, message.isRead());
```
Comment (Enuviel) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: is this test implemented here https://github.com/Leanplum/Leanplum-Android-SDK/pull/164/files#diff-d4fac6525f945e4a092843223e65630dR131 ?

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    LeanplumInboxMessage imageMessage = messagesList.get(0);
+
+    String actualUrl = imageMessage.getImageUrl().toString();
+
+    assertEquals("http://bit.ly/2GzJxxx",
+        actualUrl);
+  }
+
+  @Test
+  public void testImageFilepathIsReturnedIfPrefetchingEnabled() {
```
Comment (Enuviel) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: It can be in the testImageURL test. You do not need to create a separate test for this. Just add a comment line.

Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: We should not combine these tests, since we want our unit tests to test a single code path. This test is testing a specific behavior that is different testImageURL.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java (hunk 3 above)
Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java]: testInbox combines testing several behaviors and classes all into one giant test. This violates the Single Responsibility Principle, which all of our unit tests should respect. See this for reference: https://github.com/ghsukumar/SFDC_Best_Practices/wiki/F.I.R.S.T-Principles-of-Unit-Testing

Instead it is better to break our tests into smaller “unit tests” that cover only a single behavior. This way if we a commit breaks remove message from inbox functionality, we will fail this specific unit test and it will be easier to debug.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java (hunk 2 above)
Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java]: testInbox needs to be broken down into proper unit tests, it is a pseudo integration test. See my comment below on testing only a single behavior in each unit test.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java (hunk 6 above)
Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: No they are testing to Different behaviors.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java (hunk 5 above)
Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: I'm pulling it out of testInbox since that test is way to convoluted.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java (hunk 6 above)
Comment (Enuviel) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: "a single code path" - please check all others tests in Android SDK. Plus this is not good practice to setup all environment again if you can test it in previews one.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java (hunk 5 above)
Comment (Enuviel) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: why do we need to test twice the same?

Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: Agreed, I really want to delete the one from testInbox().

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java (hunk 6 above)
Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: 1. I think we should probably look into all the other unit tests at some point, however I think for this sprint we should move forward and increase android automation coverage. 
2. I dont really understand your second sentence, could you please clarify?

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  @Test
+  public void testOpenAction() {
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    leanplumInbox.downloadMessages();
+    LeanplumInboxMessage imageMessage = leanplumInbox.allMessages().get(1);
+    imageMessage.read();
+
+    HashMap actionName = imageMessage.getContext().objectNamed("Open action");
+
+    assertEquals(true, actionName.containsValue("Alert"));
```
Comment (Enuviel) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: actionName.containsValue must be equals "Alert"

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java (hunk 6 above)
Comment (Enuviel) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: as I mention offline - please check AbstractTest.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java [hunk 8]
```diff
@@ -48,141 +49,183 @@
...
-
-        // Validate message state.
+  LeanplumInbox leanplumInbox;
+  @Before
+  public void setUp() {
+    setupSDK(mContext, "/responses/simple_start_response.json");
+    leanplumInbox = LeanplumInbox.getInstance();
+  }
+
+  //TODO:This test needs to be refactored into smaller tests.
```
Comment (benmarten) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java]: please remove, (as discussed)

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java [hunk 9]
```diff
@@ -48,141 +49,183 @@
...
+  @Test
+  public void testInbox() throws Exception {
+
```
Comment (benmarten) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java]: please remove empty line

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java [hunk 10]
```diff
@@ -48,141 +49,183 @@
...
+    // Download messages.
+    Leanplum.getInbox().downloadMessages();
+
+    //  Remove synced callback afterwards so we don't get synced callbacks anymore.
+    Leanplum.getInbox().removeSyncedHandler(inboxSyncedCallbackFalse);
+  }
+
+  @Test
+  public void testDisablePrefetching() {
+    leanplumInbox.disableImagePrefetching();
```
Comment (benmarten) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxTest.java]: LeanplumInbox.disableImagePrefetching();

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    leanplumInbox.disableImagePrefetching();
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    leanplumInbox.downloadMessages();
+    List<LeanplumInboxMessage> messagesList = leanplumInbox.allMessages();
+    LeanplumInboxMessage imageMessage = messagesList.get(0);
+
+    String actualUrl = imageMessage.getImageUrl().toString();
+
+    assertEquals("http://bit.ly/2GzJxxx",
+        actualUrl);
```
Comment (benmarten) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: please merge with previous line if possible

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    message.read();
+
+    assertEquals(true, message.isRead());
+    assertEquals(intialUnreadCount-1, leanplumInbox.unreadCount());
+  }
+
+  @Test
+  public void testImageURL() {
+    leanplumInbox.disableImagePrefetching();
```
Comment (benmarten) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: same here (static)

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests covering Inbox Messages.
+ * @author Sayaan Saha
+ */
+
+public class LeanplumInboxMessageTest extends AbstractTest {
+  LeanplumInbox leanplumInbox;
```
Comment (benmarten) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: can this be private?

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests covering Inbox Messages.
+ * @author Sayaan Saha
+ */
+
```
Comment (benmarten) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: please remove

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
+package com.leanplum;
```
[the remark this answers is not in the document]
Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: added

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java (hunk 7 above)
Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: actionName is a hashmap which looks like {"__name__": "Alert"}, so when we call assertEquals(true, actionName.containsValue("Alert")) it is the equivalent of doing assertEquals(actionName.get("__name__"), "Alert");

i personally dont have a preference, I can switch it to the second way for better readability.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+  @Test
+  public void testOpenAction() {
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    LeanplumInbox.getInstance().downloadMessages();
+    LeanplumInboxMessage imageMessage = LeanplumInbox.getInstance().allMessages().get(1);
+    imageMessage.read();
+
+    HashMap actionName = imageMessage.getContext().objectNamed("Open action");
+
+    assertEquals("Alert", actionName.get("__name__"));
```
Comment (Enuviel) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: There you checked just that you have it in ActionContext. You need also check that when you read this message - the open action was run.

Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: The test for running the open action should be in ActionContext since it is the responsibility of ActionContext to display the message. LeanplumInboxMessage only puts messages into ActionContext, therefore the unit test should test upto the behavior of the class. I checked in ActionContext and there is a test that covers open actions in ActionContext that milos wrote.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  @Test
+  public void testIsActive() {
+    ResponseHelper.seedResponse("/responses/newsfeed_response.json");
+    LeanplumInbox.getInstance().downloadMessages();
+    LeanplumInboxMessage message = LeanplumInbox.getInstance().allMessages().get(1);
+
+    Boolean active = message.isActive();
+    assertTrue(active);
```
[the remark this answers is not in the document]
Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: Yeah that sounds good, I will add those other boundary conditions to the test.

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java (hunk 16 above)
Comment (Enuviel) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: yes, I understand this. In other words, you need to verify that  runTrackedActionNamed was called when you call 
https://github.com/Leanplum/Leanplum-Android-SDK/pull/164/files/200d38a529c046090da0e1435ba488a452be2a53#diff-8e13c25a7be11e2043a2f72ba823c3f6R138

Comment (Enuviel) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: https://github.com/Leanplum/Leanplum-Android-SDK/blob/master/AndroidSDKCore/src/main/java/com/leanplum/LeanplumInboxMessage.java#L192

Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageTest.java]: @Enuviel yes agreed, like I mentioned earlier the read method calls runTrackedActionNamed(), the functionality of that method and executing of an open action in general is the responsibility of the ActionContext class, not of LeanplumInboxMessage class. The responsibility of LeanplumInboxMessage is to correctly form an message entity and interact with the LeanplumInbox. In my test above I'm testing to ensure that the message entity has the correct open action, in this case "Alert", and testing to make sure that it can be correctly accessed by ActionContext. I don't disagree that testing whether ActionContext properly executes an open action is a good thing, however I'm saying that test should be in ActionContextTest not in LeanplumInboxMessageTest

**Comment by sayaansaha** [comment]
test this please

File: AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageWhiteboxTest.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+  public void testOpenAction() throws Exception {
+    ActionContext mock = mock(ActionContext.class);
+    Constructor<LeanplumInboxMessage> constructor = (Constructor<LeanplumInboxMessage>) LeanplumInboxMessage.class.getDeclaredConstructors()[0];
+    constructor.setAccessible(true);
+
+    LeanplumInboxMessage imageMessage = constructor.newInstance("mesageId##11", 100L,
+        200L, false, mock);
+    imageMessage.read();
+
+    verify(mock, times(1)).runTrackedActionNamed("Open action");
```
Comment (Enuviel) [inline AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageWhiteboxTest.java]: you need to call imageMessage.read(); one more time and verify that runTrackedActionNamed was called if the message was read before.

Comment (sayaansaha) [reply AndroidSDKTests/src/test/java/com/leanplum/LeanplumInboxMessageWhiteboxTest.java]: ok will do 👍
